sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -5,127 +5,217 @@ by SYCL 2020 specification: [SYCL registry][sycl-registry],
+
+Therefore, we can't have headers-only implementation and the crucial part of
+design is how to organize mapping mechanism between SYCL identifiers for
+specialization constants (`specialization_id`s) and low-level identifiers
+(numeric IDs in SPIR-V or information about corresponding kernel arguments).
+
+That mapping mechanism is particularly tricky, because of some additional
+complexity coming from SYCL 2020 specification:
+- `specialization_id` variables, which are used as specialization constant
+  identifiers (being non-type template parameters of some methods) can't be
+  forward-declared in general case (for example, if defined as `static`), which
+  means that we can't use integration header to attach some information to them
+  through some C++ templates tricks (like it is done for regular kernel
+  arguments or kernel names, for example).
+- they also can be declared as `static` or just non-`inline` `constexpr`, which
+  means that they have internal linkage and can't be referenced from other
+  translation units, which means that we can't for example create a new
+  translation unit which contains some mapping from `specialization_id` address
+  to some desired info.
+
+Based on those limitations, the following mapping design is proposed:
+- DPC++ RT uses special function:
+  ```
+  namespace detail {
+    template<auto &SpecName>
+    const char *get_spec_constant_symbolic_ID();
+  }
+  ```
+  Which is only declared, but not defined in there and used to retrieve required
+  information like numeric ID of a specialization constant.
+- Definition of that function template are provided by DPC++ FE in form of
+  _integration footer_: the compiler generates a piece of C++ code which is
+  injected at the end of a translation unit:
+  ```
+  namespace detail {
+    // assuming user defined and used the following specialization_id:
+    // constexpr specialiation_id<int> int_const;
+    // class Wrapper {
+    // public:
+    //   static constexpr specialization_id<float> float_const;
+    // };
+
+    template<>
+    const char *get_spec_constant_symbolic_ID<int_const>() {
+      return "unique_name_for_int_const";
+    }
+    template<>
+    const char *get_spec_constant_symbolic_ID<Wrapper::float_const>() {
+      return "unique_name_for_Wrapper_float_const";
+    }
+  }
+  ```
+
+  Those symbolic IDs are used to identify device image properties corresponding
+  to those specialization constants, which store additional information (like
+  numeric SPIR-V ID of a constant) needed for DPC++ RT.
+- That integration footer is automatically appended by the compiler at the end
+  of user-provided translation unit by driver.
+
+Another significant part of the design is how specialization constants support
+is emulated: as briefly mentioned before, the general approach is to transform
+specialization constants into kernel arguments. In fact all specialization
+constants used within a program a bundler together and stored into a single
```

Comment:
```suggestion
constants used within a program bundled together and stored into a single
```
?

Reply:
Indeed, fixed in 82882b9a91580d741020f2bfe694cf53fab00824